### PR TITLE
Expired Service Bindings cleanup job

### PR DESCRIPTION
**Pull request by szwedm**
**Description**

Add a job to cleanup expired Service Bindings.

**Related issue(s)**

#1347

### CODE REVIEW

File: internal/broker/client.go [hunk 1]
```diff
@@ -160,6 +178,35 @@ func (c *Client) GetInstanceRequest(instanceID string) (response *http.Response,
...
+	unbindURL, err := c.formatUnbindUrl(binding)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("sending unbind request for service binding with ID %q and instance ID: %q", binding.ID, binding.InstanceID)
+
+	emptyResponse := &apiresponses.EmptyResponse{}
+	for requestAttemptNum := 1; requestAttemptNum <= c.RequestRetries; requestAttemptNum++ {
+		if err = c.executeRequest(http.MethodDelete, unbindURL, http.StatusOK, nil, emptyResponse); err != nil {
```
Comment (jaroslaw-pieszka) [inline internal/broker/client.go:192]: How do we handle 410 response? The `Unbind` endpoint will return 410 if Service Instance had been removed before DELETE was called.

File: internal/broker/client.go [hunk 2]
```diff
@@ -245,10 +292,18 @@ func (c *Client) formatDeprovisionUrl(instance internal.Instance) (string, error
...
-	request, err := http.NewRequest(method, url, body)
+func (c *Client) formatUnbindUrl(binding internal.Binding) (string, error) {
+	if len(binding.InstanceID) == 0 {
+		return "", fmt.Errorf("empty InstanceID")
+	}
+
+	return fmt.Sprintf(unbindTmpl, c.brokerConfig.URL, instancesURL, binding.InstanceID, binding.ID), nil
+}
+
+func (c *Client) executeRequest(method, url string, expectedStatus int, requestBody io.Reader, responseBody interface{}) error {
```
Comment (jaroslaw-pieszka) [inline internal/broker/client.go:303]: As far as I remember you pass http.StatusOK in the invocation.
So http.StatusGone would we treated as not acceptable?

File: internal/servicebindingcleanup/service.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+func (s *Service) PerformCleanup() error {
+	slog.Info(fmt.Sprintf("Fetching Service Bindings with expires_at <= %q", time.Now().UTC().Truncate(time.Second).String()))
+	bindings, err := s.bindingsStorage.ListExpired()
+	if err != nil {
+		return err
+	}
+
+	slog.Info(fmt.Sprintf("Expired Service Bindings: %d", len(bindings)))
+	if s.dryRun {
```
Comment (jaroslaw-pieszka) [inline internal/servicebindingcleanup/service.go:40]: I wonder if:
```go
if !s.dryRun {
   slog.Info()
   ...
}
return nil
```
is easier to read.

File: internal/servicebindingcleanup/service_test.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+		// then
+		actualBinding, err := bindingsStorage.Get(expectedBinding.InstanceID, expectedBinding.ID)
+		require.NoError(t, err)
+		assert.Equal(t, expectedBinding.ID, actualBinding.ID)
+
+		// cleanup
+		require.NoError(t, bindingsStorage.Delete(expectedBinding.InstanceID, expectedBinding.ID))
+	})
+
+	t.Run("should unbind service bindings", func(t *testing.T) {
```
Comment (jaroslaw-pieszka) [inline internal/servicebindingcleanup/service_test.go:63]: Could we add test case to unbind expired but leave not expired?

File: internal/servicebindingcleanup/service.go (hunk 3 above)
Comment (szwedm) [reply internal/servicebindingcleanup/service.go:40]: I agree, even better if we drop the conditional clause for sending requests:
```go
if s.dryRun {
  return nil
}
slog.Info("Requesting Service Bindings removal...")
...
```
What do you think?

File: internal/servicebindingcleanup/service_test.go (hunk 4 above)
Comment (szwedm) [reply internal/servicebindingcleanup/service_test.go:63]: Yes! Good idea :)

File: internal/storage/postsql/read.go [hunk 5]
```diff
@@ -55,6 +55,23 @@ func (r readSession) ListBindings(instanceID string) ([]dbmodel.BindingDTO, erro
 	return bindings, err
 }
 
+func (r readSession) ListExpiredBindings() ([]dbmodel.BindingDTO, error) {
+	currentTime := time.Now().UTC()
+	var bindings []dbmodel.BindingDTO
+	_, err := r.session.
+		Select("*").
```
[the remark this answers is not in the document]
Comment (szwedm) [reply internal/storage/postsql/read.go:62]: I would like to improve the query to select only columns which are required for the job, that is `id`, `instance_id`, `expires_at`. The sort is redundant, we don't care about the order in which expiration requests are sent.

File: internal/broker/client.go (hunk 1 above)
Comment (szwedm) [reply internal/broker/client.go:192]: atm `executeRequest` method will throw an error and the job quits. Let me improve the method to return an error with HTTP status code from the response, so the job could log warning and act accordingly, that is continue its work when the instance or binding are gone.

File: internal/broker/client.go (hunk 2 above)
Comment (szwedm) [reply internal/broker/client.go:303]: correct, `http.StatusGone` would be unexpected. I think it's ok to return an error in such case and leave it up to the method caller how to handle the error. If the error included HTTP status code, then the caller could act accordingly.

File: internal/broker/client.go (hunk 1 above)
Comment (jaroslaw-pieszka) [reply internal/broker/client.go:192]: Discussed in direct communication.

File: internal/broker/client.go (hunk 2 above)
Comment (jaroslaw-pieszka) [reply internal/broker/client.go:303]: Discussed in direct communication.

File: docs/contributor/06-10-keb-cronjobs.md [hunk 6]
```diff
@@ -2,10 +2,11 @@
...
+| [Trial Cleanup CronJob](06-40-trial-free-cleanup-cronjobs.md)             | Causes Kyma runtime instances with the trial plan to expire 14 days after their creation.                                                                                                                   |
+| [Free Cleanup CronJob](06-40-trial-free-cleanup-cronjobs.md)              | Causes Kyma runtime instances with the free plan to expire 30 days after their creation.                                                                                                                    |
+| [Deprovision Retrigger CronJob](06-50-deprovision-retrigger-cronjob.md)   | Makes another attempt to deprovision an instance.                                                                                                                                                           |
+| [Service Binding Cleanup CronJob](06-70-deprovision-retrigger-cronjob.md) | Cleans up expired Service Bindings.                                                                                                                                                                         |
```
[the remark this answers is not in the document]
Comment (szwedm) [reply docs/contributor/06-10-keb-cronjobs.md:12]: suggestion applied.

File: docs/contributor/06-70-service-binding-cleanup-cronjob.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
+# Service Binding Cleanup CronJob
+
+Service Binding Cleanup CronJob is a Job that removes expired service bindings for SAP BTP, Kyma runtime instances.
```
[the remark this answers is not in the document]
Comment (szwedm) [reply docs/contributor/06-70-service-binding-cleanup-cronjob.md:3]: suggestion applied.

File: docs/contributor/06-70-service-binding-cleanup-cronjob.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+## Details
+
+For each expired service binding, a DELETE request is sent to Kyma Environment Broker (KEB). The request has a time limit and can be retried if it timeouts.
+
+### Dry-Run Mode
+
+If you need to test the Job, you can run it in the `dry-run` mode.
```
[the remark this answers is not in the document]
Comment (szwedm) [reply docs/contributor/06-70-service-binding-cleanup-cronjob.md:11]: suggestion applied.

File: docs/contributor/06-70-service-binding-cleanup-cronjob.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+In that mode, the Job only logs the information about the number of expired service bindings without sending DELETE requests to KEB.
```
[the remark this answers is not in the document]
Comment (szwedm) [reply docs/contributor/06-70-service-binding-cleanup-cronjob.md:12]: suggestion applied.

**Review by jaroslaw-pieszka — DISMISSED** [review]

**Review by IwonaLanger — APPROVED** [review]

**Review by jaroslaw-pieszka — APPROVED** [review]

**Pull request merged**